python/cuspatial/cuspatial/io/geoseries_reader.py

```diff
@@ -0,0 +1,284 @@
+                )
+                for interior in geometry.interiors:
+                    rings_current = offsets["polygons"]["rings"][-1]
+                    offsets["polygons"]["rings"].append(
+                        len(interior.coords) * 2 + rings_current
+                    )
+                    num_rings = num_rings + 1
+                current = offsets["polygons"]["polygons"][-1]
+                offsets["polygons"]["polygons"].append(num_rings + current)
+            elif isinstance(geometry, MultiPolygon):
+                current = offsets["polygons"]["polygons"][-1]
+                offsets["polygons"]["mpolys"].append(
+                    len(offsets["polygons"]["polygons"]) - 1
+                )
+                for poly in geometry:
+                    current = offsets["polygons"]["polygons"][-1]
+                    num_rings = 1
+                    rings_current = offsets["polygons"]["rings"][-1]
+                    offsets["polygons"]["rings"].append(
+                        len(poly.exterior.coords) * 2 + rings_current
+                    )
+                    for interior in poly.interiors:
+                        rings_current = offsets["polygons"]["rings"][-1]
+                        offsets["polygons"]["rings"].append(
+                            len(interior.coords) * 2 + rings_current
+                        )
+                        num_rings = num_rings + 1
+                    offsets["polygons"]["polygons"].append(num_rings + current)
+                offsets["polygons"]["mpolys"].append(
+                    len(offsets["polygons"]["polygons"]) - 1
+                )
+        return offsets
+
+    def _read_geometries(self, geoseries, offsets):
+        """
+        Creates a set of buffers sized to fit all of the geometries and
+        iteratively populates them with geometry coordinate values.
+
+        Parameters
+        ----------
+        geoseries : A GeoPandas GeoSeries object.
+        offsets : The set of offsets that correspond to the geoseries argument.
+        """
+        buffers = {
+            "points": cudf.Series(np.zeros(offsets["points"][-1])),
+            "multipoints": cudf.Series(np.zeros(offsets["multipoints"][-1])),
+            "lines": cudf.Series(np.zeros(offsets["lines"][-1])),
+            "polygons": {
+                "polygons": cudf.Series(
+                    np.zeros(len(offsets["polygons"]["polygons"]))
+                ),
+                "rings": cudf.Series(
+                    np.zeros(len(offsets["polygons"]["rings"]))
+                ),
+                "coords": cudf.Series(
+                    np.zeros(offsets["polygons"]["rings"][-1])
+                ),
+            },
+        }
+        read_count = {
+            "points": 0,
+            "multipoints": 0,
+            "lines": 0,
+            "polygons": 0,
+        }
+        inputs = []
+        input_types = []
+        input_lengths = []
+        for geometry in geoseries:
```

Comment:
Looping sequential over the geoseries in Python sounds like it will be very slow. Can we use something like Arrow or something similar to give us an in-memory representation that we can work with?

Reply:
Here we discuss the desired feature:
https://github.com/geopandas/geo-arrow-spec/issues

Presently geoPandas stores each geometry object as a GEOS object, which is backed by GDAL. The python API essentially mirrors the C API for them IIRC. The faster method is to move this python code into C/C++ and then push it as separate PRs for geoPandas and GEOS. This implementation took a few weeks to write and could honestly be moved into the GEOS space almost just as fast.

Reply:
ACK, lets stay with Python iteration for now then and optimize later if needed.